docs/building-modules/messages-and-queries.md

```diff
@@ -4,29 +4,51 @@ order: 3
 
-Defining `message`s is the responsibility of module developers. Typically, they are defined as protobuf messages in a `proto/` directory (see more info about [conventions and naming](../core/encoding.md#faq)). The `message`'s definition usually includes a list of parameters needed to process the message that will be provided by end-users when they want to create a new transaction containing said `message`.
+### `Msg` Services
 
-Here's an example of a protobuf message definition:
+Starting from v0.40, defining Protobuf `Msg` services is the recommended way to handle messages. A `Msg` protobuf service should be created per module, typically in `tx.proto` (see more info about [conventions and naming](../core/encoding.md#faq)). It must have an RPC service method defined for each message in the module.
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/d55c1a26657a0af937fa2273b38dcfa1bb3cff9f/proto/cosmos/gov/v1beta1/tx.proto#L15-L27
+See an example of a `Msg` service definition from `x/bank` module:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc1/proto/cosmos/bank/v1beta1/tx.proto#L10-L17
+
+For backwards compatibility with [legacy amino `Msg`s](#legacy-amino-msgs), existing `Msg` types should be used as the request parameter for `service` definitions. Newer `Msg` types which only support `service` definitions should use the more canonical `Msg...Request` names.
+
+`Msg` request types need to implement the `MsgRequest` interface which is a simplified version of the `Msg` interface described [below](#legacy-msgs) with only `ValidateBasic()` and `GetSigners()` methods.
+
+Defining such `Msg` services allow to specify return types as part of `Msg` response using the canonical `Msg...Response` names.
+
+In addition, this generates client and server code.
+`MsgServer` interface defines the server API for the `Msg` service and its implementation is described as part of the [`Msg` services](./msg-services.md) documentation.
+
+A `RegisterMsgServer` method is also generated and should be used to register the module's `MsgServer`  request types with custom type URLs in `RegisterServices` method from the [`AppModule` interface](./module-manager.md#appmodule).
+
+In order for clients (CLI and grpc-gateway) to have these URLs registered, the SDK provides the function `RegisterMsgServiceDesc(registry codectypes.InterfaceRegistry, sd *grpc.ServiceDesc)` that should be called inside module's [`RegisterInterfaces`](module-manager.md#appmodulebasic) method, using proto-generated `&_Msg_serviceDesc` as `*grpc.ServiceDesc` argument.
+`RegisterInterfaces` should be called before `MsgServiceRouter` `RegisterService(sd *grpc.ServiceDesc, handler interface{})` so that all `Msg` types are already registered by then.
+
+### Legacy Amino `Msg`s
```

Comment:
Should we just drop this section?

cc/ @amaurymartiny @robert-zaremba @clevinson @anilCSE @aaronc